cranelift/codegen/meta/src/shared/instructions.rs

```diff
@@ -4579,8 +4579,7 @@ pub(crate) fn define(
             r#"
         Atomically load from memory at `p`.
 
-        This is a polymorphic instruction that can load any value type which has a memory
-        representation.  It should only be used for integer types with 8, 16, 32 or 64 bits.
+        It should only be used for integer types with 32 or 64 bits.
```

Comment:
Please don't change this. In rustc_codegen_cranelift I use `atomic_load` to load i8 and i16 values.

Reply:
+1, it's better to be consistent with normal load instructions here and accept any type, I think.

Reply:
Okay.

Reply:
So I'm attempting to revert the IR changes, so we have the atomic_load + uextend. So during lowering, I'm trying to pattern match from the uextend, which I think works fine. But then I'm assuming that the AtomicLoad node is also being lowered so my output code ends up with two loads. Is there a way to explicitly replace a value or remove an instruction?

Reply:
Ah, interesting! So the lowering framework keeps track of reference counts for values; instructions are lowered in a backward pass where inputs are only produced if something depends on them, or if they have side-effects. However in this case the load is considered to have a side effect (it can trap). This actually needs slightly different handling; it's not legal to duplicate the load (imagine if e.g. a store interleaved).

In cases where we merge an effectful op (like a load), we need to call `sink_inst()` on the context as well; see [this comment](https://github.com/bytecodealliance/wasmtime/blob/d551997657c64ef77e6860f44241ed306787a6df/cranelift/codegen/src/machinst/lower.rs#L121-L125). An example of how this is used is [here](https://github.com/bytecodealliance/wasmtime/blob/d551997657c64ef77e6860f44241ed306787a6df/cranelift/codegen/src/isa/x64/lower.rs#L192-L198) in the x64 backend.

Sorry this is a bit subtle!

Reply:
`is_mergeable_load` returns `None` for `ty.bits() < 32`.

Reply:
That's specific to the x64 backend, because of the way that narrow loads become wider in some cases; I imagine that doesn't apply here since the widening loads don't reach beyond their width.